File: main/check-completeness-of-a-binary-tree/check-completeness-of-a-binary-tree-2.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public boolean isCompleteTree(final TreeNode root) {
        if (root == null) return true;
        
        boolean gotLeaf = false;
        final Queue<TreeNode> nodes = new ArrayDeque<>();
        
        for (nodes.add(root); !nodes.isEmpty(); ) {
            final TreeNode node = nodes.remove();
            
            if (node.left == null)
                gotLeaf = true;
            else if (gotLeaf)
                return false;
            else
                nodes.add(node.left);
            
            if (node.right == null)
                gotLeaf = true;
            else if (gotLeaf)
                return false;
            else
                nodes.add(node.right);
        }
        
        return true;
    }
}

File: main/check-completeness-of-a-binary-tree/check-completeness-of-a-binary-tree-2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    static bool isCompleteTree(const TreeNode* root) noexcept;
};

bool Solution::isCompleteTree(const TreeNode* const root) noexcept
{
    if (!root) return true;
    
    auto got_leaf = false;
    queue<const TreeNode*> q;
    
    for (q.push(root); !q.empty(); ) {
        const auto node = q.front();
        q.pop();
        
        if (!node->left)
            got_leaf = true;
        else if (got_leaf)
            return false;
        else
            q.push(node->left);
        
        if (!node->right)
            got_leaf = true;
        else if (got_leaf)
            return false;
        else
            q.push(node->right);
    }
    
    return true;
}
